social_engagement/engagement.py

```diff
@@ -346,11 +413,75 @@ def _get_request(incoming_request, params):
 
 def _get_author_of_comment(parent_id):
     comment = cc.Comment.find(parent_id)
-    if comment and hasattr(comment, 'username'):
-        return comment.username
+    if comment and hasattr(comment, 'user_id'):
+        return comment.user_id
 
 
 def _get_author_of_thread(thread_id):
     thread = cc.Thread.find(thread_id)
-    if thread and hasattr(thread, 'username'):
-        return thread.username
+    if thread and hasattr(thread, 'user_id'):
+        return thread.user_id
+
+
+def _get_thread_details_for_deletion(request, results):
+    """
+    Get details of thread and related users that are required for deletion purposes.
+    """
+    from lms.djangoapps.discussion_api.views import CommentViewSet
+    response_page = 1
+    has_results = True
+    while has_results:
```

Comment:
Usually this kind of pagination logic can be made generic. There are 2 ways I can think of:

1. Have a function that keeps on requesting data from new pages until exhausted, and then return those results so they can be operated on.
2. Have a function that accepts a function `A` which requests new data from new pages until exhausted while continuously running `A` on each set of results (thus similar to what this is doing already, but just made generic).

Either approach leads to issues though if the total number of results are _very_ large, because we have to hold this data in memory. A better approach, but which is hard to implement, is to break the processing of these results-for-deletion into separate asynchronous jobs, so that no job ends up taking up an insane amount of memory. Something like a pseudo-MapReduce.

Reply:
I've made it a bit more generic. What do you think?

Do you suggest to make a Celery task for retrieving every page from the API and then merge the results?

Reply:
Looks fine, no need for a celery task, I was just going over-kill :p